### PR TITLE
Linux/ARM: Fix +3000 bus errors of unit-test in case of O2/O3 levels

### PR DESCRIPTION
**Pull request by leemgs**
**PROBLEM**
This patch is to resolve +3000 bus errors that are reproduced whenever we use
the aggressive optimization levels of clang (issue #5844 ).

When we enable the -O3 optimization level of the clang version(from clang 3.5
to clang 3.9(snapshot)), we have always got the lots of bus errors from the
coreCLR's unit tests. Actually, we can easily monitor SIGBUS signals (e.g.,
"misaligned memory access") with /proc/cpu/alignment facility in kernel space.
Using "echo 2 > /proc/cpu/alignment" makes Linux kernel fixes the problems
but the performance of the application will be degraded.
.source: http://lxr.free-electrons.com/source/Documentation/arm/mem_alignment

**VERSION 4**
. Use 'GET_UNALIGNED_VALXXX' macros in the CoreClr infra-structure without
  any need for ifdefs.

**VERSION 3**
.Apply this PR on only Linux/ARM for different system environment (Windows).

Here is .NET CI Report on Windows: Compile Error
error C2146: syntax error: missing ';' before identifier '__unaligned_int32'
(compiling source file D:\j\workspace\checked_windo---f6dc6fe4\src\jit\alloc.cpp)
[D:\j\workspace\checked_windo---f6dc6fe4\bin\obj\Windows_NT.x64.Checked\src\jit\
crossgen\clrjit_crossgen.vcxproj] Indication 1

**VERSION 2**
.Add UNALIGNED_ARM macro for handling ARM core specific optimisation levels.
.Add RISC-based ARM core handling into the existing infra-structure of the
platform adaptation layer (PAL) for aggressive optimization cases on Linux/ARM.

**VERSION 1**
Basically, RISC-based ARM architecture requires aligned access with 4byte reads.
In order to support aggressive optimization levels such as O2/O3, let's use
attribute keyword of aligned(1) instead of using memcpy(2) in into
a properly aligned buffer or the packing attribute.

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>

### CODE REVIEW

**Comment by leemgs** [comment]
Below is unit-test result of CoreCLR. 

**Before applying the patch:**

The number of unit-test failures is 3886 (-O2 and -O3 make same result in release build mode) 

```
=======================
     Test Results (Release + O3)
=======================
# CoreCLR Bin Dir  : /unit-test/bin.coreclr.release.20160627.1.O3/Product/Linux.arm.Release
# Tests Discovered : 9873
# Passed           : 5626
# Failed           : 3886 (commit: 20160627-52570683, OOM: 1, manual kill commands: 7 )
# Skipped          : 361
=======================
102 minutes and 10 seconds taken to run CoreCLR tests.

real	103m44.120s
user	256m19.410s
sys	13m29.340s
root@target:/unit-test/coreclr.git/tests#
``` 


**After applying the patch:**

The number of unit-test failures is 47 (such as debug build mode) 
```
=======================
     Test Results (Release + O3 ; with patch ; But, needs refactoring)
=======================
# CoreCLR Bin Dir  : /unit-test/bin.coreclr.release.20160627.1.O3.attribute.1.refacto1/Product/Linux.arm.Release
# Tests Discovered : 9873
# Passed           : 9465
# Failed           : 47 (commit: 20160627-52570683, OOM: 2,  manual kill commands: 0 ) <=== N O W ! ! !
# Skipped          : 361
=======================
117 minutes and 42 seconds taken to run CoreCLR tests.

real	119m15.377s
user	145m21.515s
sys	17m23.065s

```

**Comment by leemgs** [comment]
@jkotas , Could you review and merge this PR to enable the release build mode + O3 (without O1)  on the Linux/ARM board? 

/CC: @janvorli , @benpye

File: src/pal/inc/pal.h [hunk 1]
```diff
@@ -201,7 +201,13 @@ extern "C" {
 
 #else // _MSC_VER
 
+#if defined (_M_ARM) && !defined (DEBUG)
```
Comment (janvorli) [inline src/pal/inc/pal.h]: Please use `_ARM_` instead of `_M_ARM`.
And why do you define it differently for DEBUG and non-DEBUG? It seems it would not hurt anything.

File: src/pal/inc/pal.h [hunk 2]
```diff
@@ -201,7 +201,13 @@ extern "C" {
...
 #define UNALIGNED
+#define UNALIGNED_ARM     __attribute__((aligned(1))) //for -O2 and -O3
```
Comment (janvorli) [inline src/pal/inc/pal.h]: Can you just define the UNALIGNED to __attribute__((aligned(1))) instead of introducing the UNALIGNED_ARM?

File: src/pal/inc/pal.h (hunk 1 above)
Comment (leemgs) [inline src/pal/inc/pal.h]: > Please use _ARM_ instead of _M_ARM.

Basically, I though that the valid macro is _M_ARM. I will try to change the macro from _M_ARM to _ARM. 

> And why do you define it differently for DEBUG and non-DEBUG? 

I have defined it for DEBUG (-O0) and non-DEBUG (for Checked/-O2 and Release/-O3). 

> It seems it would not hurt anything.

I can not guarantee it would not hut anything because I did have enough evaluation. I would like to remove "!defined (DEBUG)" flag after evaluating the unit-test and the aging-test enough as a next step if we need.

File: src/pal/inc/pal.h (hunk 2 above)
Comment (leemgs) [inline src/pal/inc/pal.h]: > Can you just define the UNALIGNED to __attribute__((aligned(1))) instead of introducing the UNALIGNED_ARM?

There are too many "UNALIGNED" macros in CoreCLR. It looks like the trial to modify the existing UNALIGNED statements would be a tedious change to us too. 

* For more details:  my local branch (https://github.com/leemgs/coreclr/commit/fa3cc7892bef03c9ac5355e853a692dbd645bf77)
 , I have been working the next PR to remove `UNALIGNED_ARM` in the near future.

**Comment by leemgs** [comment]
> Basically, I though that the valid macro is _M_ARM. I will try to change the macro from _M_ARM to _ARM.

The PR is Updated with squash command.

**Comment by janvorli** [comment]
@leemgs ah, I see now, the `__attribute__((aligned(1)))` cannot be used in cast, just in typedefs.
But can you please try if defining UNALIGNED to `__unaligned` like we do on windows and see if it would work too? That would be the simplest overall the code base. I've tried to do it this way on x64 ubuntu and the compiler didn't complain.

**Comment by leemgs** [comment]
> That would be the simplest overall the code base. 
> I've tried to do it this way on x64 ubuntu and the compiler didn't complain.

@janvorli , Yes. I know that. However,  I had got the `error: unterminated conditional directive`  while building CoreCLR  in case of Linux/ARM architecture.  As I shared previously, I had evaluated the effect of the below approaches (https://github.com/dotnet/coreclr/issues/5844#issuecomment-231715755).  And, some approaches could not fix all bus errors (e.g. approximately 1200 failures /3000 failures)

**Comment by jkotas** [comment]
Most of the codebase uses macros/methods like `GET_UNALIGNED_VAL16` to read unaligned values. It may be best to switch these two places in the JIT to use this macro as well to fix this problem - instead of trying to tweak the definition of `UNALIGNED`.

**Comment by leemgs** [comment]
@jkotas , According to the above CI report,  the Windows environment generates `Compile Error`. Could you share me the version of clang/llvm  that CI system uses on Windows OS to handle this case?

* http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/checked_ubuntu_flow_prtest/4220/console


* Compile Error
error C2146: syntax error: missing ';' before identifier '__unaligned_int32' (compiling source file D:\j\workspace\checked_windo---f6dc6fe4\src\jit\alloc.cpp) [D:\j\workspace\checked_windo---f6dc6fe4\bin\obj\Windows_NT.x64.Checked\src\jit\crossgen\clrjit_crossgen.vcxproj]
Indication 1

**Comment by jkotas** [comment]
Windows build is using Microsoft Visual Studio 2015 compilers.

**Comment by janvorli** [comment]
@leemgs the problem with Windows build here is that the UNALIGNED_ARM is not defined. The pal.h is never included on Windows, since there is no PAL.

**Comment by leemgs** [comment]
@jkotas , @janvorli  Thank you for sharing the information and hint about the CI build error. I will check it to resolve the build error of CI.

**Comment by leemgs** [comment]
>  Linux ARM Emulator Cross Debug Build 
> umount: /opt/linux-arm-emulator-root/dev: device is busy.

@sjsinju, @prajwal-aithal , Could you check the issue of `Linux ARM Emulator Cross ***`?  This issue results from the incorrect operation of mount/unmount.

**Comment by leemgs** [comment]
I have executed `squash`  to make one patch because all CI tests are passed. 
@jkotas , @janvorli  Could you  review and merge the updated PR to fix the unit-test failures of CoreCLR in Release Build + O3 (by default)?

File: src/jit/compiler.hpp [hunk 3]
```diff
@@ -757,13 +757,26 @@ inline
   signed __int16    getI2LittleEndian(const BYTE * ptr)
 { return * (UNALIGNED signed __int16 *)ptr; }
 
+#if defined(_ARM_) && defined(__linux__) && !defined (DEBUG)
```
Comment (janvorli) [inline src/jit/compiler.hpp]: Could you please use the GET_UNALIGNED_VAL32 and GET_UNALIGNED_VAL64 in these functions instead of this change as @jkotas has suggested before? It would make this change cleaner without any need for ifdefs.
Also, please modify all of the functions, not just the ones that were causing problems now.

Comment (leemgs) [inline src/jit/compiler.hpp]: > Could you please use the GET_UNALIGNED_VAL32 and GET_UNALIGNED_VAL64 in 
> these functions instead of this change as @jkotas has suggested before? It would make 
> this change cleaner without any need for ifdefs.

Yes. I have modified the exiting PR. I will  again send  the PR as soon as I verify the effect of  new PR with the CoreCLR's unit-test.

**Comment by leemgs** [comment]
I have sent new PR using  the GET_UNALIGNED_VALxxx. The evaluation result of unit-test will be appeared shortly.

File: src/pal/inc/pal_endian.h [hunk 4]
```diff
@@ -98,8 +98,9 @@ inline void SwapGuid(GUID *pGuid)
 #define VALPTR(x) VAL32(x)
 #endif
 
-#if defined(ALIGN_ACCESS) && !defined(_MSC_VER)
-
+#if defined(_TARGET_ARM_) && !defined(_MSC_VER)
```
Comment (leemgs) [inline src/pal/inc/pal_endian.h]: @janvorli , @jkotas  Could give me a hint about the `defined(ALIGN_ACCESS)` ?  Is `ALIGN_ACCESS` coming from ./src/inc/stdmacros.h#L131 ?  I could not know the related history why this statement was created in the past, because it was made by 'dotnet-bot'. 

```
^ef1e2ab (dotnet-bot        2015-01-30 14:14:42 -0800 103 #if defined(ALIGN_ACCESS) && !def
^ef1e2ab (dotnet-bot        2015-01-30 14:14:42 -0800 104) #ifdef __cplusplus
^ef1e2ab (dotnet-bot        2015-01-30 14:14:42 -0800 105) extern "C++" {
```

When I checked a real `ALIGN_ACCESS` definition at build time, Actually, the `ALIGN_ACCESS` (in ./src/pal/inc/pal_endian.h) macro was not defined in case of cross-compilation of CoreCLR for Linux/ARM. So, I have modified the existing `if defined(ALIGN_ACCESS) && !defined(_MSC_VER)` statement to aware aggressive optimization levels (e.g., O2/O3) in Linux/ARM environment.  Could you  review the related lines that I modified?

File: src/pal/inc/pal_endian.h [hunk 5]
```diff
@@ -98,8 +98,9 @@ inline void SwapGuid(GUID *pGuid)
...
+#define LOG2_PTRSIZE	2
```
Comment (leemgs) [inline src/pal/inc/pal_endian.h]: @jkotas , @janvorli  The line 102-103  is derived from ./src/inc/stdmacros.h#L131.  Could you review the lines are correct?

**Comment by leemgs** [comment]
> @sjsinju, @prajwal-aithal , Could you check the issue of Linux ARM Emulator Cross ***? This issue results from the incorrect operation of mount/unmount.

@sjsinju, Thanks. The ARM CI is running normally.   When this PR will be merged, Please, Append the `./common/AboveStackLimit/AboveStackLimit.exe` unit-test into https://github.com/dotnet/coreclr/blob/master/tests/testsRunningInsideARM.txt to check a bus error. 
/CC: @jyoungyun

File: src/pal/inc/pal_endian.h (hunk 5 above)
Comment (jkotas) [inline src/pal/inc/pal_endian.h]: I would keep the existing code as is, and just defined the `ALIGN_ACCESS` macro. It will make it easy to define it for new platforms as necessary. Also, this should be just `_ARM_` not `_TARGET_ARM_`.

```
#ifdef _ARM_
#define LOG2_PTRSIZE	2 
#define ALIGN_ACCESS    ((1<<LOG2_PTRSIZE)-1) 
#endif

#if defined(ALIGN_ACCESS) && !defined(_MSC_VER) 
... existing code ...
```

**Comment by jkotas** [comment]
LGTM modulo comment

File: src/pal/inc/pal_endian.h (hunk 5 above)
Comment (leemgs) [inline src/pal/inc/pal_endian.h]: @jkotas  Thanks. Applied.

**Comment by leemgs** [comment]
I re-submitted new PR (w/ squash) after applying the review of @jkotas. Below is evaluation result of the unit-test. 

* CoreCLR commit 52570683489377e50bef1edacb80b8f647223daf
* On Raspberry Pi2 Board
* Build Mode: Release build + O3 (by default)
* Before the PR: (OOM: 1, manual kill commands: 7 )
```
=======================
     Test Results 
=======================
# CoreCLR Bin Dir  : /unit-test/bin.coreclr.release.O3/Product/Linux.arm.Release
# Tests Discovered : 9873
# Passed           : 5626
# Failed           : 3886 
# Skipped          : 361
=======================
102 minutes and 10 seconds taken to run CoreCLR tests.

real	103m44.120s
user	256m19.410s
sys	13m29.340s
root@target:/unit-test/coreclr.git/tests# 
```

* After the PR: (OOM: 2, manual kill commands: 0 )
```
=======================
     Test Results
=======================
# CoreCLR Bin Dir  : /unit-test/bin.coreclr.release.O3.attribute.1.refacto1.2/Product/Linux.arm.Release
# Tests Discovered : 9873
# Passed           : 9465
# Failed           : 47
# Skipped          : 361
=======================
117 minutes and 48 seconds taken to run CoreCLR tests.

real	119m22.713s
user	145m51.545s
sys	17m8.125s
```

**Comment by jkotas** [comment]
👍

**Comment by leemgs** [comment]
> 16:41:26 Setting status of 37dc93535b8bc28c7fb4b12d1160f4faf460be93 
> to FAILURE with url http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/
> job/x86_ryujit_checked_windows_nt_prtest/2421/ and message: 'Build finished. '
> 16:41:26 Using context: Windows_NT x86 ryujit Checked Build and Test
> 16:41:26 [WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
> 16:41:27 Finished: FAILURE

@dotnet-bot test Windows_NT x86 ryujit Checked Build and Test please

**Comment by leemgs** [comment]
@jkotas , @janvorli Could you merge this PR to fix the bus errors of CoreCLR with Release Build + O3(by default)? Below  is the experimental result of CorCLR unitTest with the latest commit number.  As a next step, We will go to the -Oz (for size)  and -Ofast (for speed).

* Commit Number: c358f77a1ddcf31475450c40af64ad115259e622  (Jul-25-2016)
* UnitTest Result After the PR: (OOM: 2, manual kill commands: 1 )
```
=======================
     Test Results
=======================
# CoreCLR Bin Dir  : /unit-test/bin.coreclr.release.O3.054d133a.20160725.1/Product/Linux.arm.Release
# Tests Discovered : 9873
# Passed           : 9484
# Failed           : 28
# Skipped          : 361
=======================
164 minutes and 14 seconds taken to run CoreCLR tests.

real	165m24.788s
user	153m41.410s
sys	56m11.885s
root@target:/unit-test/coreclr.git/tests# 
```
